src/common/connection.js

```diff
@@ -95,14 +99,50 @@ class Connection extends EventEmitter {
     return this._state === WebSocket.OPEN && this._isReady;
   }
 
-  _onUnexpectedClose(resolve = function() {}, reject = function() {}) {
+  _onUnexpectedClose(resolve, reject) {
     if (this._onOpenErrorBound) {
       this._ws.removeListener('error', this._onOpenErrorBound);
       this._onOpenErrorBound = null;
     }
     this._ws = null;
     this._isReady = false;
-    this.connect().then(resolve, reject);
+    if (_.isFunction(resolve)) {
+      // connection was closed before it was properly opened, so we must return
+      // error to connect's caller
+      this.connect().then(resolve, reject);
+    } else {
+      // if first parameter ws lib sends close code,
+      // but sometimes it forgots about it, so default to 1006 - CLOSE_ABNORMAL
+      this.emit('disconnected', resolve || 1006);
```

Comment:
Can we change it so that onUnexpectedCloseBound always has resolve and reject bound to functions? In the case that they aren't necessary it could just bind them to the empty function.